### PR TITLE
Hw-dao3

### PR DESCRIPTION
**Pull request by vlykhvar**

### CODE REVIEW

File: src/main/java/mate/hwdao/Main.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    private static mate.hwdao.dao.impl.ManufactureDaoImpl manufactureDaoImpl;
+
+    public static void main(String[] args) {
+        ManufacturerService manufacturerService = (ManufacturerService) injector.getInstance(ManufacturerService.class);
+        Car car = new Car("Civic", new Manufacturer("Honda", "Japan"));
+        Driver driver = new Driver("Vasa", "0001");
+        car.getDrivers().add(driver);
+        System.out.println(car.toString());
+        manufacturerService.create(manufacturer);
+    }
```
Comment (gigbat) [inline src/main/java/mate/hwdao/Main.java]: add more calls to different methods and for tests need to output result after each

Comment (gigbat) [reply src/main/java/mate/hwdao/Main.java]: delete -> sout
update -> sout
create -> create -> sout

File: src/main/java/mate/hwdao/dao/ManufacturerDao.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+import java.util.List;
+import java.util.Optional;
+import mate.hwdao.model.Manufacturer;
+
+public interface ManufacturerDao {
+    public Manufacturer create(Manufacturer manufacturer);
+    public Optional<Manufacturer> get(Long id);
+    public List<Manufacturer> getAll();
+    public Manufacturer update(Manufacturer manufacturer);
+    public boolean delete(Long id);
```
Comment (gigbat) [inline src/main/java/mate/hwdao/dao/ManufacturerDao.java]: separate the lines and removed public

File: src/main/java/mate/hwdao/dao/impl/ManufactureDaoImpl.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+import java.util.List;
+import java.util.Optional;
+import mate.hwdao.dao.ManufacturerDao;
+import mate.hwdao.db.Storage;
+import mate.hwdao.lib.Dao;
+import mate.hwdao.model.Manufacturer;
+
+@Dao
+public class ManufactureDaoImpl implements ManufacturerDao {
+
```
Comment (gigbat) [inline src/main/java/mate/hwdao/dao/impl/ManufactureDaoImpl.java]: empty line, remove it

File: src/main/java/mate/hwdao/dao/impl/ManufactureDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    @Override
+    public Manufacturer create(Manufacturer manufacturer) {
+        Storage.listManufacturer.add(manufacturer);
+        return manufacturer;
+    }
+
+    @Override
+    public Optional<Manufacturer> get(Long id) {
+        return Optional.of(Storage.listManufacturer.stream().filter(x -> x.getId().equals(id)).findFirst().get());
```
Comment (gigbat) [inline src/main/java/mate/hwdao/dao/impl/ManufactureDaoImpl.java]:
```suggestion
        return Optional.ofNullAble(Storage.listManufacturer.stream().filter(x -> x.getId().equals(id)).findFirst().get());
```

File: src/main/java/mate/hwdao/db/Storage.java [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+import mate.hwdao.model.Car;
+import mate.hwdao.model.Driver;
+import mate.hwdao.model.Manufacturer;
+
+public class Storage {
+    public static final List<Car> listCar = new ArrayList<>();
+    public static final List<Driver> listDriver = new ArrayList<>();
```
Comment (gigbat) [inline src/main/java/mate/hwdao/db/Storage.java]: it isn't a requirement, we need to have the only a list of Manufacturer

File: src/main/java/mate/hwdao/model/Driver.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+import java.util.Objects;
+import mate.hwdao.db.Storage;
+
+public class Driver {
+    private Long id;
+    private String name;
+    private String licenseNumber;
+
+    public Driver (String name, String licenseNumber){
```
Comment (gigbat) [inline src/main/java/mate/hwdao/model/Driver.java]: check for null

File: src/main/java/mate/hwdao/model/Car.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+import java.util.List;
+import mate.hwdao.db.Storage;
+
+public class Car {
+    private Long id;
+    private String model;
+    private Manufacturer manufacturer;
+    private List<Driver> drivers;
+
+    public Car(String model, Manufacturer manufacturer) {
```
Comment (gigbat) [inline src/main/java/mate/hwdao/model/Car.java:12]: check for null and into the methods set()

File: src/main/java/mate/hwdao/model/Manufacturer.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    private final Long id;
+    private String name;
+    private String country;
+
+    public Manufacturer(){
+        id = setId();
+        this.name = "name";
+        this.country = "country";
+    }
+    public Manufacturer(String name, String country) {
```
Comment (gigbat) [inline src/main/java/mate/hwdao/model/Manufacturer.java:10]: same

File: src/main/java/mate/hwdao/service/ManufacturerService.java [hunk 9]
```diff
@@ -0,0 +1,13 @@
+package mate.hwdao.service;
+
+import java.util.List;
+import mate.hwdao.model.Manufacturer;
+
+public interface ManufacturerService {
+
```
Comment (gigbat) [inline src/main/java/mate/hwdao/service/ManufacturerService.java:8]: empty line

File: src/main/java/mate/hwdao/service/ManufacturerService.java [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+    public Manufacturer create(Manufacturer manufacturer);
+    public Manufacturer get(Long id);
+    public List<Manufacturer> getAll();
+    public Manufacturer update(Manufacturer manufacturer);
+    public boolean delete(Long id);
+}
```
Comment (gigbat) [inline src/main/java/mate/hwdao/service/ManufacturerService.java]: separate the lines

**Review by gigbat — CHANGES_REQUESTED** [review]
Good job 👍

**Review by gigbat — APPROVED** [review]

File: pom.xml [hunk 11]
```diff
@@ -9,8 +9,15 @@
...
     </properties>
-
+    <!-- https://mvnrepository.com/artifact/org.apache.maven.plugins/maven-clean-plugin -->
+    <dependencies>
+    <dependency>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-clean-plugin</artifactId>
+        <version>3.1.0</version>
+    </dependency>
+    </dependencies>
```
Comment (RomanDubovskyi) [inline pom.xml:28]: missing empty line at the end of the file

File: src/main/java/mate/hwdao/dao/impl/ManufactureDaoImpl.java [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+public class ManufactureDaoImpl implements ManufacturerDao {
+    @Override
+    public Manufacturer create(Manufacturer manufacturer) {
+        Storage.listManufacturer.add(manufacturer);
+        return manufacturer;
+    }
+
+    @Override
+    public Optional<Manufacturer> get(Long id) {
+        return Optional.ofNullable(Storage.listManufacturer.stream().filter(x -> x.getId().equals(id)).findFirst().get());
```
Comment (RomanDubovskyi) [inline src/main/java/mate/hwdao/dao/impl/ManufactureDaoImpl.java]: this line seems to be longer than allowed, try to run and fix checkstyle

File: src/main/java/mate/hwdao/dao/impl/ManufactureDaoImpl.java [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+    public List<Manufacturer> getAll() {
+        return Storage.listManufacturer;
+    }
+
+    @Override
+    public Manufacturer update(Manufacturer manufacturer) {
+        for (Manufacturer manufacturer1 : Storage.listManufacturer) {
+            if (manufacturer.getId().equals(manufacturer.getId())) {
+                manufacturer1 = manufacturer;
+                return manufacturer;
```
Comment (RomanDubovskyi) [inline src/main/java/mate/hwdao/dao/impl/ManufactureDaoImpl.java]: this solution won't work if you create a new object with the same id, use for i loop instead method set of the list

File: src/main/java/mate/hwdao/dao/impl/ManufactureDaoImpl.java [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+            }
+        }
+        return null;
```
Comment (RomanDubovskyi) [inline src/main/java/mate/hwdao/dao/impl/ManufactureDaoImpl.java]: returning null is a bad practice, better to throw exception

File: src/main/java/mate/hwdao/dao/impl/ManufactureDaoImpl.java [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    @Override
+    public boolean delete(Long id) {
+        return Storage.listManufacturer.remove(Storage.listManufacturer.stream().filter(x -> x.getId().equals(id)).findFirst().get());
```
Comment (RomanDubovskyi) [inline src/main/java/mate/hwdao/dao/impl/ManufactureDaoImpl.java]: try to use removeIf here instead, will make the code shorter

File: src/main/java/mate/hwdao/db/Storage.java (hunk 5 above)
Comment (RomanDubovskyi) [reply src/main/java/mate/hwdao/db/Storage.java]: \+
that's the code from next hws

File: src/main/java/mate/hwdao/db/Storage.java [hunk 16]
```diff
@@ -0,0 +1,14 @@
+package mate.hwdao.db;
+
+
+import java.util.ArrayList;
+import java.util.List;
+import mate.hwdao.model.Car;
+import mate.hwdao.model.Driver;
+import mate.hwdao.model.Manufacturer;
+
+public class Storage {
```
Comment (RomanDubovskyi) [inline src/main/java/mate/hwdao/db/Storage.java:7]: where do you store ids counter and set it to model?

**Review by RomanDubovskyi — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 11 above)
Comment (vlykhvar) [reply pom.xml:28]: added

File: src/main/java/mate/hwdao/dao/impl/ManufactureDaoImpl.java (hunk 15 above)
Comment (vlykhvar) [reply src/main/java/mate/hwdao/dao/impl/ManufactureDaoImpl.java]: done

File: src/main/java/mate/hwdao/dao/impl/ManufactureDaoImpl.java (hunk 14 above)
Comment (vlykhvar) [reply src/main/java/mate/hwdao/dao/impl/ManufactureDaoImpl.java]: done

File: pom.xml [hunk 17]
```diff
@@ -9,8 +9,59 @@
...
+            <configuration>
+                <source>${jdk.version}</source>
+                <target>${jdk.version}</target>
+                <encoding>${project.build.sourceEncoding}</encoding>
+            </configuration>
+        </plugin>
+    </plugins>
+</pluginManagement>
+</build>
+        </project>
```
Comment (RomanDubovskyi) [inline pom.xml]:
```suggestion
</project>

```

File: src/main/java/mate/hwdao/db/Storage.java [hunk 18]
```diff
@@ -0,0 +1,9 @@
+package mate.hwdao.db;
+
+import java.util.ArrayList;
+import java.util.List;
+import mate.hwdao.model.Manufacturer;
+
+public class Storage {
```
Comment (RomanDubovskyi) [inline src/main/java/mate/hwdao/db/Storage.java:7]:
```suggestion
public class Storage {
private static Long manufacturerId = 0L;
```

File: src/main/java/mate/hwdao/db/Storage.java [hunk 19]
```diff
@@ -0,0 +1,9 @@
...
+    public static final List<Manufacturer> listManufacturer = new ArrayList<>();
```
Comment (RomanDubovskyi) [inline src/main/java/mate/hwdao/db/Storage.java:8]: add method addManufacturer that will set id and increment it

File: src/main/java/mate/hwdao/service/ManufacturerServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+
+@Service
+public class ManufacturerServiceImpl implements ManufacturerService {
+    @Inject
+    private ManufacturerDao manufacturerDao;
+    private long countId = 0;
+
+    @Override
+    public Manufacturer create(Manufacturer manufacturer) {
+        manufacturer.setId(++countId);
```
Comment (RomanDubovskyi) [inline src/main/java/mate/hwdao/service/ManufacturerServiceImpl.java]: it should be done in the Storage (db level)

**Review by RomanDubovskyi — CHANGES_REQUESTED** [review]

File: src/main/java/mate/hwdao/service/ManufacturerServiceImpl.java [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+    @Override
+    public Manufacturer create(Manufacturer manufacturer) {
+        manufacturerDao.create(manufacturer);
+        return manufacturer;
+    }
+
+    @Override
+    public Manufacturer get(Long id) {
+        return manufacturerDao.get(id)
+               .orElseThrow(() -> new RuntimeException("Can't find " + id));
```
Comment (RomanDubovskyi) [inline src/main/java/mate/hwdao/service/ManufacturerServiceImpl.java:23]:
```suggestion
               .orElseThrow(() -> new RuntimeException("Can't find manufacturer with id " + id));
```

**Review by RomanDubovskyi — APPROVED** [review]

**Pull request merged**